pkgs/build-support/trivial-builders/default.nix

```diff
@@ -591,6 +591,43 @@ rec {
     ln -s ${lib.getBin drv}/bin $out/bin
   '';
 
+  /*
+    Adds a hook to the nativeBuildInputs of a derivation which will provide an
+    SSH session via tmate to the build environment, for interactive
+    debugging
+
+    Example:
+
+    # Spawn a tmate session after the `buildPhase` of the `hello` derivation
+    addTmateBreakpoint (pkgs.hello.overrideAttrs { buildPhase = "exit 1;"; })
+  */
+  addTmateBreakpoint = let
+    warningMessage = "addTmateBreakpoint in use, ignore the sha256 warning and do not leak these build logs, otherwise unauthorized ssh access to the build sandbox may occur";
+    hook = lib.warn warningMessage (makeSetupHook {} (writeTextFile {
+      name = "hook.sh";
+      text = ''
+        breakpointHookTmate() {
+          local red='\033[0;31m'
+          local no_color='\033[0m'
+
+          echo -e "''${red}build failed in ''${curPhase} with exit code ''${exitCode}''${no_color}"
+          echo -e "''${red}### WARNING ###''${no_color}"
+          echo -e "''${red}${warningMessage}''${no_color}"
+
+          printf "To attach using tmate:\n\n"
+          ${util-linux}/bin/script -qefc "${tmate}/bin/tmate -F"
```

Comment:
Maybe factor these two lines out so you can put an explicit breakpoint call anywhere in the derivation build script?

Reply:
Currently, to call the hook any where you want, you just have to place an `exit 1;` anywhere. For me, the point of the hook is to drop you in where a derivation fails. Are you proposing not doing that and instead making it a function you call from in the drv phase, so rather than `exit 1` to call the hook? Not sure what is meant by "factor these two lines out" otherwise, can you explain more?